File: client/src/models/packets/packet-server-scene-data.js
import PacketLabels from '../../constants/packet-labels'

export default class PacketServerSceneData {

  constructor(packet) {
    this.label = PacketLabels.SERVER_SCENE_DATA
    this.data = packet.data || null
    this.sceneKey = packet.sceneKey || null
  }

  receive(sceneKey, callback) {
    if (this.sceneKey !== sceneKey) {
      return
    }

    if (typeof callback === 'function') {
      callback(this)
    }
  }

}

File: server/models/packet_server_scene_data.go
package models

import (
	"encoding/json"
	"github.com/pkg/errors"
	"unlock/constants"
)

type PacketServerSceneData struct {
	Packet
	Data     interface{}        `json:"data"`
	SceneKey constants.SceneKey `json:"sceneKey"`
}

func NewPacketServerSceneData(data interface{}, sceneKey constants.SceneKey) *PacketServerSceneData {
	return &PacketServerSceneData{
		Packet: Packet{
			Label: constants.PacketServerSceneData,
		},
		Data:     data,
		SceneKey: sceneKey,
	}
}

func (packet *PacketServerSceneData) Send(lobby *Lobby) (err error) {
	payload, err := json.Marshal(packet)
	if err != nil {
		return errors.WithStack(err)
	}

	lobby.Broadcast <- payload

	return lobby.PushToFirestore()
}
